File: GuardianMonsters/battle/src/de/limbusdev/guardianmonsters/battle/model/ObservableQueue.java
package de.limbusdev.guardianmonsters.battle.model;

import com.badlogic.gdx.utils.Array;

import java.util.Comparator;

/**
 * Created by georg on 04.12.16.
 */

public class ObservableQueue<T> extends Array<T> implements ObservableList<T> {

    private Array<ListObserver> observers;

    @Override
    public void add(T value) {
        super.add(value);
    }

    @Override
    public void insert(int index, T value) {
        super.insert(index, value);
    }

    @Override
    public boolean removeValue(T value, boolean identity) {
        return super.removeValue(value, identity);
    }

    @Override
    public T removeIndex(int index) {
        return super.removeIndex(index);
    }

    @Override
    public void removeRange(int start, int end) {
        super.removeRange(start, end);
    }

    @Override
    public T pop() {
        return super.pop();
    }

    @Override
    public void clear() {
        super.clear();
    }

    @Override
    public void sort() {
        super.sort();
    }

    @Override
    public void sort(Comparator<? super T> comparator) {
        super.sort(comparator);
    }

    @Override
    public void reverse() {
        super.reverse();
    }

    @Override
    public void shuffle() {
        super.shuffle();
    }

    @Override
    public void addObserver(ListObserver o) {
        if(observers == null)
            observers = new Array<ListObserver>();
        observers.add(o);
    }

    @Override
    public void clearObservers() {
        if(observers != null)
            observers.clear();
    }

    @Override
    public void notifyObserversRemove(T item) {
        for(ListObserver lo : observers) {
            lo.getInformedAboutRemoval(item);
        }
    }

    @Override
    public void notifyObserversAdd(T item) {

    }

    @Override
    public void notifyObserversSort() {

    }

}

File: GuardianMonsters/battle/src/de/limbusdev/guardianmonsters/battle/model/ObservableQueue.kt
package de.limbusdev.guardianmonsters.battle.model

import com.badlogic.gdx.utils.Array

import java.util.Comparator

/**
 * @author Georg Eckert 2019
 */

class ObservableQueue<T> : Array<T>(), ObservableList<T>
{
    private var observers: Array<ListObserver<T>> = Array()

    override fun add(value: T)
    {
        super.add(value)
    }

    override fun insert(index: Int, value: T)
    {
        super.insert(index, value)
    }

    override fun removeValue(value: T?, identity: Boolean): Boolean
    {
        return super.removeValue(value, identity)
    }

    override fun removeIndex(index: Int): T
    {
        return super.removeIndex(index)
    }

    override fun removeRange(start: Int, end: Int)
    {
        super.removeRange(start, end)
    }

    override fun pop(): T
    {
        return super.pop()
    }

    override fun clear()
    {
        super.clear()
    }

    override fun sort()
    {
        super.sort()
    }

    override fun sort(comparator: Comparator<in T>)
    {
        super.sort(comparator)
    }

    override fun reverse()
    {
        super.reverse()
    }

    override fun shuffle() {
        super.shuffle()
    }

    override fun addObserver(o: ListObserver<T>)
    {
        observers.add(o)
    }

    override fun clearObservers()
    {
        observers.clear()
    }

    override fun notifyObserversRemove(item: T)
    {
        for (lo in observers)
        {
            lo.getInformedAboutRemoval(item)
        }
    }

    override fun notifyObserversAdd(item: T)
    {

    }

    override fun notifyObserversSort()
    {

    }

}
